Collect all file patterns when running multiple rake test tasks

domcleal opened the pull request:
Fixes #27801

domcleal committed:
Collect all file patterns when running multiple rake test tasks

Replaces the rake_patterns instance variable with simple require, as
`autorun` will run tests from all eagerly required test files.

Fixes #27801

arthurnn commented:
Thanks @domcleal for the bug fix.. Do you mind adding a changelog entry for this fix, and rebasing your PR against latest master?

kaspth commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -60,7 +60,7 @@ def self.plugin_rails_options(opts, options)
   # as the patterns would also contain the other Rake tasks.
   def self.rake_run(patterns) # :nodoc:
-    @rake_patterns = patterns
+    (@rake_patterns ||= []).push(*patterns)
```
Won't this negate what the comment says above? In other words, do we need `rake_run` anymore? It might have become obsolete after moving to `autorun`.

domcleal replied:
The comment is still valid with `@rake_patterns` - it's targeted at fixing `rake db:migrate test` rather than two test tasks. Without it, the plugin would run tests listed in the argument list, which is `['db:migrate' ,'test']` - instead `rake_run` stores the paths from test:\* tasks so running test + non-test tasks works.

This was originally added in df744b5, but missing a test to cover it and the conditional above, so I added one in b5a9428.

I think you're right that `rake_run` (or rather `@rake_patterns`) is a bit obsolete now that `autorun` is used. The test files can be required immediately and `autorun` will handle it. I've removed `@rake_patterns` in a12803d, but kept `rake_run` as it makes the tasks in railties/lib/rails/test_unit/testing.rake a little less repetitive.

domcleal commented:
@arthurnn thanks, I've rebased the PR, added two commits in response to @kaspth's comment (improve test coverage, refactor away this instance variable) and added the CHANGELOG entry.

arthurnn commented:
Aren't we going to require_files twice now? https://github.com/domcleal/rails/blob/a12803dc17417437ddbbe93bc798d6b779363331/railties/lib/rails/test_unit/minitest_plugin.rb#L90

domcleal commented:
@arthurnn thanks, you're right that it was called twice. `options[:pattern]` would be empty though for rake runs, but I've updated the last commit to skip the step when called via rake.

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -536,6 +536,21 @@ def test_rake_passes_multiple_TESTOPTS_to_minitest
+
+    def test_rake_db_and_test_tasks_parses_args_correctly
+      create_test_file :models, "account"
+      output = Dir.chdir(app_path) { `bin/rake db:migrate test:models TESTOPTS='-v' && echo ".tables" | rails dbconsole` }
```
Won't the piping into `dbconsole` make the test stall?

domcleal replied:
The test doesn't hang if that's what you mean, but it may take a bit longer to run. I wanted to check that the db:migrate task ran too, and this is the method used by previous tests.

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -536,6 +536,21 @@ def test_rake_passes_multiple_TESTOPTS_to_minitest
+      create_test_file :models, "account"
+      output = Dir.chdir(app_path) { `bin/rake db:migrate test:models TESTOPTS='-v' && echo ".tables" | rails dbconsole` }
+      assert_match "AccountTest#test_truth", output, "passing test:models should run test/models/"
+      assert_match "ar_internal_metadata", output, "tables should be dumped"
```
Let's kill the custom messages on `assert_match` here and above. Think the intention is sufficiently clear in the test code itself.

kaspth commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -86,8 +85,9 @@ def run(args = [])
-    # If run via `ruby` we've been passed the files to run directly.
-    unless run_via[:ruby]
+    # If run via `ruby` we've been passed the files to run directly, or if run
+    # via `rake` then they have already been included.
```
`included` is a new term, let's say `eagerly required` to stick with the established terminology.

kaspth commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -59,19 +59,18 @@ def self.plugin_rails_options(opts, options)
   def self.rake_run(patterns) # :nodoc:
-    @rake_patterns = patterns
+    ::Rails::TestRequirer.require_files patterns
+    run_via[:rake] = true
```
Think it makes more sense to assign `run_via` as the first step and then require the files.

kaspth commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -59,19 +59,18 @@ def self.plugin_rails_options(opts, options)
-  # as the patterns would also contain the other Rake tasks.
   def self.rake_run(patterns) # :nodoc:
-    @rake_patterns = patterns
+    ::Rails::TestRequirer.require_files patterns
```
Add parens around here: `::Rails::TestRequirer.require_files(patterns)`

kaspth commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Fix running multiple test\* tasks specified in a single `rake` command
```
Don't understand what `test\*` is supposed to mean.

I'd write:

```
*   Fix running multiple tests in one `rake` command.

    E.g. `bin/rake test:models test:controllers`

# …
```

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -536,6 +536,21 @@ def test_rake_passes_multiple_TESTOPTS_to_minitest
+    def test_rake_runs_multiple_test_tasks
+      create_test_file :models, "account"
+      create_test_file :controllers, "accountcontroller"
+      output = Dir.chdir(app_path) { `bin/rake test:models test:controllers TESTOPTS='-v'` }
```
Why do we pass in `TESTOPTS` here and below?

domcleal replied:
It prints the names of the tests, so it can assert the expected tests ran using `output`.

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -536,6 +536,21 @@ def test_rake_passes_multiple_TESTOPTS_to_minitest
 
+    def test_rake_runs_multiple_test_tasks
+      create_test_file :models, "account"
+      create_test_file :controllers, "accountcontroller"
```
Let's follow Rails' naming conventions for good measure: `"accounts_controller"` :)

kaspth commented:
Squash your commits down to 1 once the latest review has been addressed. Thanks for fixing this, digging it! 🤘

domcleal commented:
Thank you for the detailed review and comments @kaspth. I've fixed the style comments and squashed the commits.

kaspth commented:
Awesome, thanks!

domcleal commented:
Thanks @kaspth. Would you mind backporting it to 5-0-stable, since it's a regression vs 4.2?

kaspth commented:
Apologies for not keeping you in the loop, I needed https://github.com/rails/rails/pull/27941 to be merged first.

Backported to 5-0-stable at e79ffaf...88b335e

domcleal commented:
No worries, thanks @kaspth.

kaspth commented:
Sounds good 👍
